### ISSUE TITLE
No way to update the Auth user object with cognito updates when using refreshSession in React app

### ISSUE THREAD

**Issue by stuartlangridge** [aws-amplify/amplify-js#11088]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Authentication

### Amplify Categories

auth

### Environment information

<details>

```
# Put output below this line

  System:
    OS: Linux 5.15 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (4) x64 Intel(R) Core(TM) i5-4670 CPU @ 3.40GHz
    Memory: 20.34 GB / 31.29 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 18.15.0 - /usr/bin/node
    npm: 9.5.0 - /usr/bin/npm
  Browsers:
    Chrome: 111.0.5563.64
    Chromium: 111.0.5563.64
    Firefox: 110.0
  npmPackages:
    @aws-amplify/ui-react: ^4.3.8 => 4.3.8 
    @aws-amplify/ui-react-internal:  undefined ()
    @testing-library/jest-dom: ^5.16.5 => 5.16.5 
    @testing-library/react: ^13.4.0 => 13.4.0 
    @testing-library/user-event: ^13.5.0 => 13.5.0 
    aws-amplify: ^5.0.14 => 5.0.14 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    react-router-dom: ^6.8.1 => 6.8.1 
    react-scripts: 5.0.1 => 5.0.1 
    web-vitals: ^2.1.4 => 2.1.4 
  npmGlobalPackages:
    @aws-amplify/cli: 10.7.3

```

</details>

### Describe the bug

When an Amplify React app is created and uses auth, it is passed a `user` object. The details in that `user` object (and in particular Cognito custom variables) are cached at login. There seems to be no way to cause that user object to update from Cognito if the values stored in Cognito have changed since logging in, without the user logging out and in again. (`user.refreshSession` returns what would be the new content of the user object, but doesn't update the actual user object, and that user object can't be sensibly replaced.)

### Expected behavior

There should be some documented way to update the user object.

### Reproduction steps

```javascript
function App({ signOut, user }) {
  // here, we want to wait ten seconds (this is standing in for
  // the user navigating away to a payment provider and then
  // coming back again, with Cognito data having changed)
  // and then update the user object.

  // this needs to happen in a useEffect block because otherwise
  // it runs many times because it re-triggers component update
  useEffect(() => {
    const refreshTheUserObject = async () => {
      // how do we tell the user object to update?
      // we can use user.refreshSession (is this documented?)
      // but that only returns a new view of what the user object looks like
      user.clearCachedUserData();
      user.refreshSession(user.signInUserSession.refreshToken, async (err, data) => {
        // here, "data" is a "new" user object, but we can't overwrite the passed
        // user object because we're in a useEffect block, and there's no
        // user.updateTheUserObjectInPlace() method or similar documented
      })
    }
    setTimeout(refreshTheUserObject, 10000);
  })
}
```

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

**Comment by tannerabread** [comment]
This seems similar to #2987 except that one was talking about user groups not updating until the next session.

@stuartlangridge just to confirm what is not updating, but you are not seeing your [custom attributes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-custom-attributes) update or did I misunderstand and you are talking about some different custom variables?

Also just confirming that you get the updates on a new session and cache isn't breaking that?

**Comment by stuartlangridge** [comment]
> Also just confirming that you get the updates on a new session and cache isn't breaking that?
Yup. If I log out and then log in again, the user object that's passed to my App has the updated Cognito custom attributes in it. (I meant custom attributes yes, thank you for the correction!)

**Comment by tannerabread** [comment]
After speaking with the team there is not currently a way to do this with the library so we will mark it as a feature-request 

Thank you for your input and bringing this to our attention!

**Comment by tannerabread** [comment]
Hi @stuartlangridge I may have spoken too quickly yesterday saying this wasn't possible.

You can get the updated custom_attributes through the `currentAuthenticatedUser` method if you pass the `bypassCache` flag. Would something like the following work for your use case?
```js
Auth.currentAuthenticatedUser({bypassCache: true})
  .then(async user => {
    setUser(user);
  })
  .catch(err => console.log(err));
```

Let me know if this isn't what you are trying to do, but when I switched the flag to `false`, it no longer received the updates to custom attributes from Cognito, when the flag is true I can get the updates

**Comment by stuartlangridge** [comment]
> You can get the updated custom_attributes through the `currentAuthenticatedUser` method if you pass the `bypassCache` flag. Would something like the following work for your use case?

This works great! I didn't know about the `setUser` function, or `bypassCache`: did I miss them in the documentation? But this is the solution I was looking for, and does what we need, so thank you!

**Comment by tannerabread** [comment]
`setUser` was actually from my own project I had tried it on where I was keeping it in state, so that part shouldn't be necessary

We do have [`bypassCache` in the docs](https://docs.amplify.aws/lib/auth/manageusers/q/platform/js/#retrieve-current-authenticated-user) but it may not be the most obvious. If this was all you needed though, and it was just too hidden in the docs to find, we would appreciate any feedback or suggestions on how this could have been more obvious!
